File: templates/problems/08_tries/02_word_dictionary/word_dictionary_test.go
package word_dictionary

import "testing"

func TestWordDictionary(t *testing.T) {
	tests := []struct {
		name     string
		adds     []string
		searches []struct{ word string; want bool }
	}{
		{
			name: "basic with wildcards",
			adds: []string{"bad", "dad", "mad"},
			searches: []struct{ word string; want bool }{
				{"pad", false},
				{"bad", true},
				{".ad", true},
				{"b..", true},
				{"...", true},
				{"....", false},
			},
		},
		{
			name: "exact match only",
			adds: []string{"hello"},
			searches: []struct{ word string; want bool }{
				{"hello", true},
				{"hell", false},
				{"helloo", false},
			},
		},
		{
			name: "all wildcards",
			adds: []string{"ab", "cd"},
			searches: []struct{ word string; want bool }{
				{"..", true},
				{".", false},
				{"...", false},
			},
		},
		{
			name: "single char",
			adds: []string{"a"},
			searches: []struct{ word string; want bool }{
				{".", true},
				{"a", true},
				{"..", false},
			},
		},
		{
			name: "mixed wildcards",
			adds: []string{"apple", "ample"},
			searches: []struct{ word string; want bool }{
				{"a.ple", true},
				{"a..le", true},
				{"a...e", true},
				{"a....", true},
				{"b....", false},
			},
		},
		{
			name: "no words added",
			adds: []string{},
			searches: []struct{ word string; want bool }{
				{"a", false},
				{".", false},
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			wd := NewWordDictionary()
			for _, w := range tt.adds {
				wd.AddWord(w)
			}
			for _, s := range tt.searches {
				if got := wd.Search(s.word); got != s.want {
					t.Errorf("Search(%q) = %v, want %v", s.word, got, s.want)
				}
			}
		})
	}
}

File: templates/problems/08_tries/02_word_dictionary/typescript/word_dictionary.test.ts
import { describe, it, expect } from "vitest";
import { WordDictionary } from "./word_dictionary";

describe("WordDictionary", () => {
    it("basic with wildcards", () => {
        const wd = new WordDictionary();
        for (const w of ["bad", "dad", "mad"]) {
            wd.addWord(w);
        }
        expect(wd.search("pad")).toBe(false);
        expect(wd.search("bad")).toBe(true);
        expect(wd.search(".ad")).toBe(true);
        expect(wd.search("b..")).toBe(true);
        expect(wd.search("...")).toBe(true);
        expect(wd.search("....")).toBe(false);
    });

    it("exact match only", () => {
        const wd = new WordDictionary();
        wd.addWord("hello");
        expect(wd.search("hello")).toBe(true);
        expect(wd.search("hell")).toBe(false);
        expect(wd.search("helloo")).toBe(false);
    });

    it("all wildcards", () => {
        const wd = new WordDictionary();
        for (const w of ["ab", "cd"]) {
            wd.addWord(w);
        }
        expect(wd.search("..")).toBe(true);
        expect(wd.search(".")).toBe(false);
        expect(wd.search("...")).toBe(false);
    });

    it("single char", () => {
        const wd = new WordDictionary();
        wd.addWord("a");
        expect(wd.search(".")).toBe(true);
        expect(wd.search("a")).toBe(true);
        expect(wd.search("..")).toBe(false);
    });

    it("mixed wildcards", () => {
        const wd = new WordDictionary();
        for (const w of ["apple", "ample"]) {
            wd.addWord(w);
        }
        expect(wd.search("a.ple")).toBe(true);
        expect(wd.search("a..le")).toBe(true);
        expect(wd.search("a...e")).toBe(true);
        expect(wd.search("a....")).toBe(true);
        expect(wd.search("b....")).toBe(false);
    });

    it("no words added", () => {
        const wd = new WordDictionary();
        expect(wd.search("a")).toBe(false);
        expect(wd.search(".")).toBe(false);
    });
});
